plaso/storage/redis/merge_reader.py

```diff
@@ -0,0 +1,292 @@
+        task_identifier=task.identifier)
+    self._store.Open(redis_client=redis_client)
+    self._event_data_identifier_mappings = {}
+    self._add_container_type_methods = {}
+    self._active_extra_containers = []
+
+    # Create a runtime lookup table for the add container type method. This
+    # prevents having to create a series of if-else checks for container types.
+    # The table is generated at runtime as there are no forward function
+    # declarations in Python.
+    for container_type, method_name in self._ADD_CONTAINER_TYPE_METHODS.items():
+      method = getattr(self, method_name, None)
+      if not method:
+        raise RuntimeError(
+            'Add method missing for container type: {0:s}'.format(
+                container_type))
+
+      self._add_container_type_methods[container_type] = method
+
+  def _AddAnalysisReport(self, analysis_report):
+    """Adds an analysis report.
+
+    Args:
+      analysis_report (AnalysisReport): analysis report.
+    """
+    self._storage_writer.AddAnalysisReport(analysis_report)
+
+  def _AddError(self, error):
+    """Adds an error.
+
+    Args:
+      error (ExtractionError): error.
+    """
+    self._storage_writer.AddError(error)
+
+  def _AddEventSource(self, event_source):
+    """Adds an event source.
+
+    Args:
+      event_source (EventSource): event source.F
+    """
+    self._storage_writer.AddEventSource(event_source)
+
+  def _AddEventTag(self, event_tag):
+    """Adds an event tag.
+
+    Args:
+      event_tag (EventTag): event tag.
+    """
+    self._storage_writer.AddEventTag(event_tag)
+
+  def _AddEvent(self, event):
+    """Adds an event.
+
+    Args:
+      event (EventObject): event.
+    """
+    if hasattr(event, 'event_data_row_identifier'):
+      event_data_identifier = identifiers.SQLTableIdentifier(
+          self._CONTAINER_TYPE_EVENT_DATA,
+          event.event_data_row_identifier)
+      lookup_key = event_data_identifier.CopyToString()
+
+      event_data_identifier = self._event_data_identifier_mappings[lookup_key]
+      event.SetEventDataIdentifier(event_data_identifier)
+
+    # TODO: add event identifier mappings for event tags.
+
+    self._storage_writer.AddEvent(event)
+
+  def _AddEventData(self, event_data):
+    """Adds event data.
+
+    Args:
+      event_data (EventData): event data.
+    """
+    identifier = event_data.GetIdentifier()
+    lookup_key = identifier.CopyToString()
+
+    self._storage_writer.AddEventData(event_data)
+
+    post_write_identifier = event_data.GetIdentifier()
+    self._event_data_identifier_mappings[lookup_key] = post_write_identifier
+
+  def _PrepareForNextContainerType(self):
+    """Prepares for the next container type.
+
+    This method prepares the task storage for merging the next container type.
+    It sets the active container type, its add method and active cursor
+    accordingly.
+    """
+    self._active_container_type = self._container_types.pop(0)
```

Comment:
what if `self._container_types` is empty?

Reply:
For reference: https://github.com/log2timeline/plaso/pull/2799/files#diff-4b06666ecd79eed7b1e1511a572aa258R266

Reply:
Ack and that seems to be the only call, and there is a check there. Let's add this to our notes to re-evaluate changing in the future.